profiling/src/io.rs

```diff
@@ -0,0 +1,683 @@
+                if ((*dyn_iter).d_un.d_ptr as usize) < ((*info).dlpi_addr as usize) {
+                    rel_plt = ((*info).dlpi_addr as usize + (*dyn_iter).d_un.d_ptr as usize)
+                        as *mut Elf64_Rela;
+                } else {
+                    rel_plt = (*dyn_iter).d_un.d_ptr as *mut Elf64_Rela;
+                }
+            }
+            DT_PLTRELSZ => {
+                // Size of the PLT relocation entries
+                rel_plt_size = (*dyn_iter).d_un.d_val as usize;
+            }
+            DT_SYMTAB => {
+                // The symbol table
+                if ((*dyn_iter).d_un.d_ptr as usize) < ((*info).dlpi_addr as usize) {
+                    symtab = ((*info).dlpi_addr as usize + (*dyn_iter).d_un.d_ptr as usize)
+                        as *mut Elf64_Sym;
+                } else {
+                    symtab = (*dyn_iter).d_un.d_ptr as *mut Elf64_Sym;
+                }
+            }
+            DT_STRTAB => {
+                // The string table for the symbol names
+                if ((*dyn_iter).d_un.d_ptr as usize) < ((*info).dlpi_addr as usize) {
+                    strtab = ((*info).dlpi_addr as usize + (*dyn_iter).d_un.d_ptr as usize)
+                        as *const c_char;
+                } else {
+                    strtab = (*dyn_iter).d_un.d_ptr as *const c_char;
+                }
+            }
+            _ => {}
+        }
+        dyn_iter = dyn_iter.offset(1);
+    }
+
+    if rel_plt.is_null() || rel_plt_size == 0 || symtab.is_null() || strtab.is_null() {
+        trace!("Failed to locate required ELF sections (`DT_JMPREL`, `DT_PLTRELSZ`, `DT_SYMTAB` and `DT_STRTAB`)");
+        return;
+    }
+
+    let num_relocs = rel_plt_size / std::mem::size_of::<Elf64_Rela>();
+
+    // For each symbol we want to overwrite (from `overwrites`), we scan the relocation entries.
+    // Once the matching symbol name is found, patch its GOT entry to point to our new function.
+    for overwrite in &mut *overwrites {
+        for i in 0..num_relocs {
+            let rel = rel_plt.add(i);
+            let r_type = elf64_r_type((*rel).r_info);
+
+            // Only handle JUMP_SLOT relocations
+            if r_type != R_AARCH64_JUMP_SLOT && r_type != R_X86_64_JUMP_SLOT {
+                continue;
+            }
+
+            // Get the symbol index for this relocation, then the symbol struct
+            let sym_index = elf64_r_sym((*rel).r_info) as usize;
+            let sym = symtab.add(sym_index);
+
+            // Access the symbol name via the string table
+            let name_offset = (*sym).st_name as isize;
+            let name_ptr = strtab.offset(name_offset);
+            let name = CStr::from_ptr(name_ptr).to_str().unwrap_or("");
+
+            if name == overwrite.symbol_name {
+                // Calculate the GOT entry address
+                let got_entry =
+                    ((*info).dlpi_addr as usize + (*rel).r_offset as usize) as *mut *mut ();
```

Comment:
Is this guaranteed to be a `sizeof(void *)` aligned address? Otherwise you might crash when attempting to write across the page boundary. Please check, and if yes, add a comment.

Reply:
Jup, it is aligned 🎉 
Added a comment and source with 1447603